File: app/src/main/java/Controller/Cart/CartRecyclerViewAdapter.java
package Controller.Cart;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.graphics.BitmapFactory;
import android.text.InputType;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.elsafa.R;
import com.example.elsafa.ui.Cart.CartFragment;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;

import java.util.List;
import java.util.Map;

import Model.Item;
import Model.Product.Product;

public class CartRecyclerViewAdapter extends RecyclerView.Adapter<CartRecyclerViewAdapter.ViewHolder> implements OnGetProductsListener {

    private final CartFragment cartFragment;
    private final FirebaseFirestore fStore;
    private final FirebaseAuth auth;
    private final List<Item> list;
    private final Context context;
    private final ProductsPresenter presenter;


    public CartRecyclerViewAdapter(List<Item> list, Context context, CartFragment cartFragment) {
        this.list = list;
        this.context = context;
        this.cartFragment = cartFragment;
        fStore = FirebaseFirestore.getInstance();
        auth = FirebaseAuth.getInstance();

        presenter = new ProductsPresenter(this);
    }

    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.cart_item_card_view, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        presenter.getProduct(holder, list.get(position), position);

        setRemoveListener(holder, position);
    }

    private void addPrice(float v) {
        cartFragment.editTotalBill(v);
    }

    private void setRemoveListener(ViewHolder holder, int position) {
        holder.remove.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                removeCartFromProduct(position);
            }
        });
    }

    private void setEditListener(ViewHolder holder, int position, float price) {
        holder.edit.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                editClicked(holder, position, price);
            }
        });
    }

    private void editClicked(ViewHolder holder, int position, float price) {
        AlertDialog.Builder builder = new AlertDialog.Builder(context);

        EditText editText = new EditText(context);
        editText.setInputType(InputType.TYPE_CLASS_NUMBER);

        builder.setView(editText);
        builder.setTitle("Enter Quantity");
        builder.setPositiveButton("Change", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                int quantity = Integer.parseInt(editText.getText().toString());
                holder.quantity.setText(String.valueOf(quantity));
                holder.total_price.setText(quantity * price + " EGP");
                setQuantity(position, quantity);
            }
        });
        builder.setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                dialog.cancel();
            }
        });

        builder.show();
    }

    private void setQuantity(int position, int quantity) {
        fStore.collection("Users")
                .document(auth.getUid())
                .collection("Cart")
                .document(list.get(position).getCategoryName() + list.get(position).getProductId())
                .update("quantity", quantity);
    }

    private void removeCartFromProduct(int position) {
        fStore.collection("Categories")
                .document(list.get(position).getCategoryName())
                .collection("Products")
                .document(list.get(position).getProductId())
                .get()
                .addOnSuccessListener(new OnSuccessListener<DocumentSnapshot>() {
                    @Override
                    public void onSuccess(DocumentSnapshot ds) {
                        Map<String, Integer> map = (Map<String, Integer>) ds.get("carts");
                        map.remove(auth.getUid());
                        fStore.collection("Categories")
                                .document(list.get(position).getCategoryName())
                                .collection("Products")
                                .document(list.get(position).getProductId())
                                .update("carts", map);

                        removeProductFromCart(position);
                    }
                });
    }

    private void removeProductFromBill(Double price) {
        cartFragment.editTotalBill(-1 * price);
    }

    private void removeProductFromCart(int position) {
        fStore.collection("Users")
                .document(auth.getUid())
                .collection("Cart")
                .document(list.get(position).getCategoryName() + list.get(position).getProductId())
                .delete();

        cartFragment.updateView();
    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    @Override
    public void onSuccess(ViewHolder holder, Product product, int position) {
        holder.imageView.setImageBitmap(BitmapFactory.decodeByteArray(product.getMainPic().toBytes()
                , 0, product.getMainPic().toBytes().length));
        holder.title.setText(product.getTitle());

        int quantity = list.get(position).getQuantity();
        holder.quantity.setText(String.valueOf(quantity));

        float totalPrice = (product.getPrice() * quantity);
        holder.total_price.setText(totalPrice + " EGP");

        setEditListener(holder, position, product.getPrice());
    }

    @Override
    public void onFailed(Exception e) {

    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        ImageView imageView;
        TextView title, total_price, quantity;
        TextView edit, remove;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            imageView = itemView.findViewById(R.id.item_image);
            title = itemView.findViewById(R.id.title);
            total_price = itemView.findViewById(R.id.total_price);
            quantity = itemView.findViewById(R.id.quantity);

            edit = itemView.findViewById(R.id.edit);
            remove = itemView.findViewById(R.id.remove);
        }
    }
}

File: app/src/main/java/com/whiteside/cafe/adapter/CartRecyclerViewAdapter.kt
package com.whiteside.cafe.adapter

import android.app.AlertDialog
import android.text.InputType
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.EditText
import android.widget.Toast
import androidx.recyclerview.widget.RecyclerView
import com.whiteside.cafe.common.BaseListener
import com.whiteside.cafe.common.BaseViewHolder
import com.whiteside.cafe.databinding.CartItemCardViewBinding
import com.whiteside.cafe.model.Item
import com.whiteside.cafe.model.Product
import com.whiteside.cafe.ui.cart.CartPresenter
import com.whiteside.cafe.ui.product.ProductPresenter
import com.whiteside.cafe.utils.BlobBitmap
import javax.inject.Inject

class CartRecyclerViewAdapter @Inject constructor(
    var cartPresenter: CartPresenter,
    var productPresenter: ProductPresenter
) : RecyclerView.Adapter<CartRecyclerViewAdapter.ViewViewHolder>() {

    var list: ArrayList<Item> = ArrayList()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewViewHolder {
        return ViewViewHolder(
            CartItemCardViewBinding.inflate(LayoutInflater.from(parent.context))
        )
    }

    override fun onBindViewHolder(holder: ViewViewHolder, position: Int) {
        productPresenter.getProduct(list[position].categoryName, list[position].productId) {
            holder.updateView(it, list[position])
        }
    }

    override fun getItemCount(): Int {
        return list.size
    }


    inner class ViewViewHolder(val bind: CartItemCardViewBinding) : BaseViewHolder(bind) {
        fun updateView(product: Product, item: Item) {
            bind.itemImage.setImageBitmap(BlobBitmap.convertBlobToBitmap(product.productPics[0]))
            bind.title.text = product.title
            bind.quantity.text = item.quantity.toString()
            bind.totalPrice.text = "${product.price * item.quantity!!}"

            bind.edit.setOnClickListener {
                showDialog(item)
            }

            bind.remove.setOnClickListener {
                cartPresenter.removeItem(product,
                    object : BaseListener<Product> {
                        override fun onSuccess(result: Product) {
                            handleSuccess()
                            bind.root.removeAllViews()
                            notifyDataSetChanged()
                            Toast.makeText(
                                bind.root.context,
                                "Item removed from cart",
                                Toast.LENGTH_SHORT
                            ).show()
                        }

                        override fun onFailed(exception: Exception) {
                            handleFailure(exception)
                        }

                        override fun onLoading() {
                            handleLoading()
                        }
                    })
            }
        }

        private fun showDialog(item: Item) {
            val editText = EditText(bind.root.context)
            editText.inputType = InputType.TYPE_CLASS_NUMBER

            AlertDialog.Builder(bind.root.context)
                .setView(editText)
                .setTitle("Enter Quantity")
                .setPositiveButton("Change") { _d, _ ->

                    cartPresenter.updateQuantity(item, editText.text.toString().toInt()) {
                        bind.quantity.text = editText.text.toString()
                        _d.cancel()
                    }

                }.setNegativeButton("Cancel") { _d, _ ->
                    _d.cancel()
                }.show()
        }
    }
}